internal/pluginframework/tfschema/struct_to_schema_test.go

```diff
@@ -0,0 +1,152 @@
+}
+
+type TestListTfSdk struct {
+	Repeated []types.Int64 `tfsdk:"repeated" tf:"optional"`
+}
+
+type TestMapTfSdk struct {
+	Attributes map[string]types.String `tfsdk:"attributes" tf:"optional"`
+}
+
+type TestNestedListTfSdk struct {
+	NestedList []DummyNested `tfsdk:"nested_list" tf:"optional"`
+}
+
+type DummyNested struct {
+	Name    types.String `tfsdk:"name" tf:"optional"`
+	Enabled types.Bool   `tfsdk:"enabled" tf:"optional"`
+}
+
+type TestNestedMapTfSdk struct {
+	NestedMap map[string]DummyNested `tfsdk:"nested_map" tf:"optional"`
+}
+
+type TestPointerTfSdk struct {
+	Nested *DummyNested `tfsdk:"nested" tf:"optional"`
+}
+
+var tests = []struct {
+	name       string
+	testStruct any
+}{
+	{
+		"string conversion",
+		TestStringTfSdk{Description: types.StringValue("abc")},
+	},
+	{
+		"bool conversion",
+		TestBoolTfSdk{Enabled: types.BoolValue(true)},
+	},
+	{
+		"int conversion",
+		TestIntTfSdk{Workers: types.Int64Value(1)},
+	},
+	{
+		"float conversion",
+		TestFloatTfSdk{Float: types.Float64Value(1.1)},
+	},
+	{
+		"list conversion",
+		TestListTfSdk{Repeated: []types.Int64{types.Int64Value(12), types.Int64Value(34)}},
+	},
+	{
+		"map conversion",
+		TestMapTfSdk{Attributes: map[string]types.String{"key": types.StringValue("value")}},
+	},
+	{
+		"nested list conversion",
+		TestNestedListTfSdk{NestedList: []DummyNested{
+			{
+				Name:    types.StringValue("abc"),
+				Enabled: types.BoolValue(true),
+			},
+			{
+				Name:    types.StringValue("def"),
+				Enabled: types.BoolValue(false),
+			},
+		}},
+	},
+	{
+		"nested map conversion",
+		TestNestedMapTfSdk{NestedMap: map[string]DummyNested{
+			"key1": {
+				Name:    types.StringValue("abc"),
+				Enabled: types.BoolValue(true),
+			},
+			"key2": {
+				Name:    types.StringValue("def"),
+				Enabled: types.BoolValue(false),
+			},
+		}},
+	},
+	{
+		"pointer conversion",
+		TestPointerTfSdk{
+			&DummyNested{
+				Name:    types.StringValue("def"),
+				Enabled: types.BoolValue(true),
+			},
+		},
+	},
+}
+
+// StructToSchemaConversionTestCase runs a single test case to verify StructToSchema works for both data source and resource.
+func StructToSchemaConversionTestCase(t *testing.T, description string, testStruct any) {
+	scm := ResourceStructToSchema(testStruct, nil)
+	state := tfsdk.State{
+		Schema: scm,
+	}
+	// Assert we can properly set the state, this means the schema and the struct are consistent.
+	assert.True(t, !state.Set(context.Background(), testStruct).HasError(), fmt.Sprintf("ResourceStructToSchema - %s", description))
+
+	data_scm := DataSourceStructToSchema(testStruct, nil)
+	data_state := tfsdk.State{
+		Schema: data_scm,
+	}
+	// Assert we can properly set the state, this means the schema and the struct are consistent.
+	assert.True(t, !data_state.Set(context.Background(), testStruct).HasError(), fmt.Sprintf("DataSourceStructToSchema - %s", description))
+}
+
+func TestStructToSchemaConversion(t *testing.T) {
```

Comment:
Awesome. It would also be great to have some negative test cases: where is this expected to fail? E.g.

```go
func testStructToSchemaPanics(t *testing.T, struct any, expectedError string) {
  defer func() {
    err := recover()
    if err == nil {
      t.Fail(fmt.Sprintf("did not fail with error %q", expectedError))
    }
    if !strings.Contains(err.Error(), expectedError) {
      t.Fail(fmt.Sprintf("error %w did not include expected error message %q", err, expectedError))
    }
  }
  ResourceStructToSchema(struct)
}
```
We can add test cases for unsupported paths.